Mathlib/Algebra/Module/Presentation/Cokernel.lean

```diff
@@ -0,0 +1,148 @@
+and one for each generator of `M₁`. -/
+@[simps]
+def cokernelRelations : Relations A where
+  G := pres₂.G
+  R := Sum pres₂.R ι
+  relation x := match x with
+    | .inl r => pres₂.relation r
+    | .inr i => data.lift i
+
+/-- The obvious solution in `M₂ ⧸ LinearMap.range f` to the equations in
+`pres₂.cokernelRelations data`. -/
+@[simps]
+def cokernelSolution :
+    (pres₂.cokernelRelations data).Solution (M₂ ⧸ LinearMap.range f) where
+  var g := Submodule.mkQ _ (pres₂.var g)
+  linearCombination_var_relation := by
+    intro x
+    erw [← Finsupp.apply_linearCombination]
+    obtain (r | i) := x
+    · erw [pres₂.linearCombination_var_relation]
+      dsimp
+    · erw [data.π_lift]
+      simp
+
+variable (hg₁ : Submodule.span A (Set.range g₁) = ⊤)
+
+namespace cokernelSolution
+
+/-- The cokernel can be defined by generators and relations. -/
+noncomputable def isPresentationCore :
+    Relations.Solution.IsPresentationCore.{w}
+      (pres₂.cokernelSolution data) where
+  desc s := (LinearMap.range f).liftQ (pres₂.desc
+    { var := s.var
+      linearCombination_var_relation :=
+        fun r ↦ s.linearCombination_var_relation (.inl r) }) (by
+          rw [LinearMap.range_eq_map, ← hg₁, Submodule.map_span, Submodule.span_le,
+            Set.image_subset_iff]
+          rintro _ ⟨i, rfl⟩
+          rw [Set.mem_preimage, SetLike.mem_coe, LinearMap.mem_ker, ← data.π_lift,
+            Relations.Solution.IsPresentation.π_desc_apply]
+          exact s.linearCombination_var_relation (.inr i))
+  postcomp_desc s := by aesop
+  postcomp_injective h := by
+    ext : 1
+    apply pres₂.toIsPresentation.postcomp_injective
+    ext g
+    exact Relations.Solution.congr_var h g
+
+include hg₁ in
+lemma isPresentation : (pres₂.cokernelSolution data).IsPresentation :=
+  (isPresentationCore pres₂ data hg₁).isPresentation
+
+end cokernelSolution
+
+/-- The presentation of the cokernel of a linear map `f : M₁ →ₗ[A] M₂` that is obtained
+from a presentation `pres₂` of `M₂`, a choice of generators `g₁ : ι → M₁` of `M₁`,
+and an additional data in `pres₂.CokernelData f g₁`. -/
+@[simps!]
+def cokernel : Presentation A (M₂ ⧸ LinearMap.range f) :=
+  ofIsPresentation (cokernelSolution.isPresentation pres₂ data hg₁)
+
+end Cokernel
+
+/-- Given an exact sequence of `A`-modules `M₁ → M₂ → M₃ → 0`, this is the presentation
+of `M₃` that is obtained from a presentation `pres₂` of `M₂`, a choice of generators
+`g₁ : ι → M₁` of `M₁`, and an additional data in a `Presentation.CokernelData` structure. -/
+@[simps!]
+noncomputable def ofExact {f : M₁ →ₗ[A] M₂} {g : M₂ →ₗ[A] M₃}
+    (pres₂ : Presentation.{w₂₀, w₂₁} A M₂) {ι : Type w₁} {g₁ : ι → M₁}
+    (data : pres₂.CokernelData f g₁)
+    (hfg : Function.Exact f g) (hg : Function.Surjective g)
+    (hg₁ : Submodule.span A (Set.range g₁) = ⊤) :
+    Presentation A M₃ :=
+  ofIsPresentation ((pres₂.cokernel data hg₁).ofLinearEquiv
+    (hfg.linearEquivOfSurjective hg))
```

Comment:
This is `Presentation.ofLinearEquiv` from #18359 right?
Also `IsPresentation.ofLinearEquiv` should be `IsPresentation.of_linearEquiv`.

Reply:
Thanks very much! I have fixed these inconsistencies in these PRs.